### ISSUE TITLE
`DefaultSavePlanner._validate_global_plan` is costly slow when FSDP size is extremely large (quadratic)

### ISSUE THREAD

**Issue by HAOCHENYE** [pytorch/pytorch#163548]
## Problem
[`_validate_global_plan`](https://github.com/pytorch/pytorch/blob/bf28990c3dd4d8bfa3eeb36a69cf53ed5f9e583e/torch/distributed/checkpoint/default_planner.py#L624) has O(n²) complexity that becomes unusably slow for modern LLMs. With 1024 tensors and `fsdp_size` 1024, validation takes 200+ seconds. if `fsdp_size` is 2048,  this can take 10+ minutes, making it unusable in production.

## Reproduction
```python
from torch.distributed.checkpoint.default_planner import DefaultSavePlanner
from torch.distributed.tensor import distribute_tensor, Shard
from torch.distributed.device_mesh import init_device_mesh
import torch
import torch.distributed as dist
import time

pg = torch.distributed.init_process_group(backend="gloo")
world_size = dist.get_world_size(pg)
rank = dist.get_rank(pg)
device_mesh = init_device_mesh("cpu", (world_size,))

fully_tensor = [torch.ones(1024, 1) for _ in range(1024)]
sharded_tensor = [distribute_tensor(tensor=t, device_mesh=device_mesh, placements=[Shard(0)]) for t in fully_tensor]
state_dict = {str(key): value for key, value in enumerate(sharded_tensor)}

planner = DefaultSavePlanner()
planner.set_up_planner(state_dict=state_dict, is_coordinator=rank==0)
local_plan = planner.create_local_plan()
gather_objs = [None] * world_size

dist.gather_object(obj=local_plan, object_gather_list=gather_objs if rank == 0 else None, dst=0, group=pg)

if rank == 0:
    start = time.time()
    all_local_plans, global_metadata = planner.create_global_plan(gather_objs)
    end = time.time()
    print(f"Create global planner cost {end - start}s")

dist.barrier()
```

Run with: `torchrun --nproc-per-node 1024 script.py`

**Result**: Takes 200+ seconds for 1024 tensors

cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @pragupta @ezyang @msaroufim @dcci

**Comment by ezyang** [comment]
Here is what ChatGPT thinking thinks should be done, but I have not read over it. The safest strategy is to NOT change semantics of the function and I do not know how to evaluate if semantics changing is safe. Please do not add 3p checks.

----

You’re paying the quadratic tax in the inner “check every later chunk for overlap” loop. Below are several ways to keep the exact validation semantics (same return + same warnings for every overlapping pair) but make the common case (no/few overlaps) much faster.

⸻

1) Sweep-line over one axis + active set (O(n log n + k))

Idea: Sort chunks by their start on the first dimension; sweep forward keeping an active set of chunks whose first-dim interval still intersects the current position. Only compare the current chunk to the active set, additionally filtering by the remaining dimensions.
	•	n = #chunks for a given tensor, k = #overlapping pairs (often 0).
	•	Worst case still Ω(k) comparisons (can’t avoid if you must log each pair), but no-overlap plans become O(n log n) instead of O(n²).

Sketch:

```
from bisect import insort
from math import prod

def _interval(box, d):  # [start, end) on dimension d
    start = box.offsets[d]
    return start, start + box.sizes[d]

def _overlaps_all_dims(a, b, ndim):
    # Your existing _check_box_overlap is fine; inline for speed if needed
    for d in range(ndim):
        a0, a1 = _interval(a, d)
        b0, b1 = _interval(b, d)
        if a1 <= b0 or b1 <= a0:  # no overlap on dim d
            return False
    return True

def _overlap_pairs_sweepline(chunks, ndim, first_dim=0):
    # Build events on first_dim
    events = []  # (coord, type, idx) type: 0 = start, 1 = end to make starts processed first at ties
    for i, ch in enumerate(chunks):
        s, e = _interval(ch, first_dim)
        events.append((s, 0, i))
        events.append((e, 1, i))
    events.sort()

    active = []         # store indices of active chunks
    active_ends = []    # parallel list sorted by end coordinate on first_dim for fast pruning

    def first_end(i):
        _, e = _interval(chunks[i], first_dim)
        return e

    for coord, typ, i in events:
        if typ == 1:  # end: remove i from active
            # lazy removal: leave tombstones and periodically compact, or keep a dict {i:pos}
            continue
        # prune any already-ended items from the front of active lists
        # (optional if you maintain a heap/ordered structure)
        j = 0
        while j < len(active) and first_end(active[j]) <= coord:
            j += 1
        if j > 0:
            del active[:j]
            del active_ends[:j]

        # check overlaps against active set (all still overlap in first_dim)
        for idx in active:
            if _overlaps_all_dims(chunks[i], chunks[idx], ndim):
                yield (idx, i)

        # insert current into active ordered by end
        end = first_end(i)
        pos = 0
        while pos < len(active_ends) and active_ends[pos] <= end:
            pos += 1
        active.insert(pos, i)
        active_ends.insert(pos, end)
```

Drop this into your validator: iterate for (i,j) in _overlap_pairs_sweepline(value.chunks, len(value.size)): and log exactly like today. This preserves semantics (every overlapping pair still logged) while making the typical tiling case near-linear.

Notes
	•	You can speed the “active set” with heapq (plus an index-guard) or sortedcontainers (3rd-party) for clean O(log n) inserts/prunes.
	•	If chunk sizes vary wildly, pick the sweep axis with largest spread / lowest overlap density (e.g., the longest dimension) for a smaller active set.

⸻

2) Spatial index (R-tree) for exact pair enumeration (≈ O(n log n + k))

If you’re okay with a dependency, use an R-tree: insert each AABB once, query intersections in ~log time, log every hit pair.

```
# pip install rtree
from rtree import index

def _overlap_pairs_rtree(chunks):
    # Build 3D..kD boxes as flatten (x0,x1,y0,y1,...) in the index.
    p = index.Property()
    idx = index.Index(properties=p)
    def bounds(ch):
        # (min_x, max_x, min_y, max_y, ...) in the order expected by rtree
        out = []
        for d in range(len(ch.sizes)):
            s = ch.offsets[d]
            e = s + ch.sizes[d]
            out.extend([s, e])
        return tuple(out)

    for i, ch in enumerate(chunks):
        idx.insert(i, bounds(ch))

    for i, ch in enumerate(chunks):
        for j in idx.intersection(bounds(ch)):
            if j <= i:       # avoid dup/self
                continue
            # Double-check exact overlap across all dims (rtree is bbox-based)
            if _check_box_overlap(ch, chunks[j]):
                yield (i, j)
```

Same semantics, great average performance, minimal code.

⸻

3) Coordinate-compression + difference array (detect any overlap + coverage)

If your chunks lie on a coarse grid (very common), compress all unique boundaries per dimension (≤ 2n per dim), and operate on the induced cells.
	•	Add each chunk as +1 to a D-dimensional difference array at its compressed corners (inclusion–exclusion).
	•	Prefix-sum along each dimension → each cell’s coverage count.
	•	If any cell’s count > 1 ⇒ overlap exists; also compute union volume by summing count>0 cells × cell volume.
	•	Compare union volume to tensor_volume for the full-cover check.

This makes no pair enumeration (so it’s much faster) but changes logging semantics (you’d no longer print the specific (chunk0, chunk1) pairs). If precise per-pair warnings are not strictly required, this is the fastest exact method for grid-like chunking: typically O(n log n + ∏ m_d) where m_d = #unique boundaries on dim d, often close to linear in n for regular tilings.

⸻

4) Hash bucketing (uniform grid)

Partition space into buckets (per-dim size ~ typical chunk size). For each chunk, add it to all buckets it intersects; only compare chunks within the same bucket.
	•	Guarantees no misses if you add to every overlapped bucket.
	•	Average time ~O(n) when chunks are near a grid; worst-case degrades if some chunks span many buckets.

⸻

5) Micro-optimizations (cheap wins)
	•	Precompute values locally to avoid global lookups in hot loops:

check_bounds = _check_box_bounds
check_overlap = _check_box_overlap
prod_ = math.prod


	•	Use math.prod instead of reduce(operator.mul, ...).
	•	Compute tensor_volume = math.prod(value.size) once before the loop.
	•	If len(global_plan) <= 1, you can skip the volume-equals-tensor check (you already do).

⸻

Drop-in update example (sweep-line)

Here’s how you’d minimally change your validator to remove the quadratic inner loop while preserving all warnings:

```
import math

def _validate_global_plan(global_plan: list[SavePlan], metadata: Metadata) -> bool:
    all_good = True
    for key, value in metadata.state_dict_metadata.items():
        if isinstance(value, BytesStorageMetadata) or len(value.size) == 0:
            continue

        # bounds + volume
        chunks_volume = 0
        for ch in value.chunks:
            if not _check_box_bounds(value.size, ch):
                logger.warning(
                    "key:%s has out of bounds chunk: tensor-size:%s chunk:%s",
                    key, value.size, ch
                )
                all_good = False
            chunks_volume += math.prod(ch.sizes)

        # overlaps via sweep-line
        for i, j in _overlap_pairs_sweepline(value.chunks, len(value.size)):
            logger.warning("key:%s has overlapping chunks: %s %s", key, value.chunks[i], value.chunks[j])
            all_good = False

        # full coverage (same as today)
        tensor_volume = math.prod(value.size)
        if len(global_plan) > 1 and chunks_volume != tensor_volume:
            logger.warning(
                "key:%s invalid fill tensor-volume: %s chunks-volume: %s",
                key, tensor_volume, chunks_volume
            )
            all_good = False

    return all_good
```

⸻

Which should you pick?
	•	Expect no overlaps (typical tiling): Sweep-line (1) — trivial to maintain, no extra deps, big win.
	•	General large plans, want robust speed: R-tree (2) — smallest code, consistently fast.
	•	Strict grid, don’t need per-pair logs: Coordinate compression (3) — fastest exact check and gives you union volume “for free”.
	•	Heuristic speedup with little code: Hash bucketing (4) — good in practice, simple to reason about.

If you want, I can tailor one of these to your actual chunk shape (attribute names/types) and drop in fully tested code.
